R/derive_extreme_event.R

```diff
@@ -106,61 +106,181 @@
+#'    and finally, because we wish to compare all records with each other, we
+#'    select `join_type = "all"`.
+#'  - Since we are now passing an `event_joined` object to
+#'    `derive_extreme_event()`, we also need to specify the `source_datasets`
+#'    object so as to indicate which dataset we wish to use for the join.
+#'
+#' @code
+#' adqs2 <- tribble(
+#'    ~USUBJID, ~PARAMCD,     ~AVALC, ~ADY,
+#'    "4",      "WAKE UP",    "N",    1,
+#'    "4",      "NO SLEEP",   "Y",    2,
+#'    "4",      "NO SLEEP",   "Y",    3,
+#'    "5",      "NO SLEEP",   "N",    1,
+#'    "5",      "NO SLEEP",   "Y",    2,
+#'    "5",      "WAKE UP 3X", "Y",    3,
+#'    "5",      "NO SLEEP",   "N",    4
+#' )
+#'
+#' derive_extreme_event(
+#'   adqs2,
+#'   by_vars = exprs(USUBJID),
+#'   events = list(
+#'     event_joined(
+#'       dataset_name = "adqs2",
+#'       join_vars = exprs(AVALC, PARAMCD, ADY),
+#'       join_type = "all",
+#'       condition = PARAMCD == "NO SLEEP" & AVALC == "Y" &
+#'         PARAMCD.join == "NO SLEEP" & AVALC.join == "Y" &
+#'         ADY == ADY.join + 1,
+#'       set_values_to = exprs(AVALC = "No sleep two nights in a row", AVAL = 0)
+#'     ),
+#'     event(
+#'       condition = PARAMCD == "NO SLEEP" & AVALC == "Y",
+#'       set_values_to = exprs(AVALC = "No sleep", AVAL = 1)
+#'     ),
+#'     event(
+#'       condition = PARAMCD == "WAKE UP 3X" & AVALC == "Y",
+#'       set_values_to = exprs(AVALC = "Waking up three times", AVAL = 2)
 #'     ),
 #'     event(
 #'       condition = all(AVALC == "N"),
 #'       set_values_to = exprs(
-#'         AVALC = "No sleeping problems", AVAL = 4
+#'         AVALC = "No sleeping problems", AVAL = 3
 #'       )
 #'     ),
 #'     event(
 #'       condition = TRUE,
 #'       set_values_to = exprs(AVALC = "Missing", AVAL = 99)
 #'     )
 #'   ),
+#'   source_datasets = list(adqs2 = adqs2),
 #'   tmp_event_nr_var = event_nr,
 #'   order = exprs(event_nr, desc(ADY)),
 #'   mode = "first",
 #'   set_values_to = exprs(
 #'     PARAMCD = "WSP",
-#'     PARAM = "Worst Sleeping Problems"
-#'   )
+#'     PARAM = "Worst Sleeping Problem"
+#'   ),
+#'   keep_source_vars = exprs(everything())
 #' )
 #'
-#' # Use different mode by event
+#' @caption Specifying different arguments across `event()` object
+#' @info Here we consider a Hy's law use case. We are interested in
+#'   knowing whether a subject's Alkaline Phosphatase has ever not
+#'   been below twice the upper limit of normal range. If so, i.e. if
+#'   `CRIT1FL` is `NA`, we are interested in the record for the first
+#'    time this occurs, and if not, we wish to retain the last record.
+#'
+#'  - In first `event()`, since we simply seek the first time that
+#'    `CRIT1FL` is `NA`, it's enough to specify the `condition`.
+#'  - In the second `event()`, we select the last record among the
+#'    full set of records where `CRIT1FL = "Y"` by additionally
+#'     specifying`mode = "last"` within the `event()`.
+#'
```

Comment:
I think we should add a bullet to call out usage of `keep_source_vars = exprs(AVISITN)` here instead of the usual `exprs(everything)`, which makes sense here as you wouldn't want `CRIT1FL` populated for the new records.

Reply:
great point, will add.